iwyu_ast_util.cc

```diff
@@ -546,32 +560,32 @@ class TypeEnumerator : public RecursiveASTVisitor<TypeEnumerator> {
 
   // --- Methods on RecursiveASTVisitor
   bool TraverseType(QualType type) {
-    TraverseArgumentsOfSugaredTemplates(type);
+    TraverseTypeHelper(type);
     return Base::TraverseType(type);
   }
 
   bool TraverseTypeLoc(TypeLoc type_loc) {
-    TraverseArgumentsOfSugaredTemplates(type_loc.getType());
+    TraverseTypeHelper(type_loc.getType());
     return Base::TraverseTypeLoc(type_loc);
   }
 
-  bool VisitType(Type* type) {
-    seen_types_.insert(type);
-    return true;
-  }
-
  private:
-  // Clang doesn't traverse underlying type for most of the sugar types.
-  // If a TemplateSpecializationType occurs in a sugaring chain, traverse its
-  // arguments explicitly, otherwise they aren't put into resugar_map.
-  void TraverseArgumentsOfSugaredTemplates(QualType type) {
-    if (type.isNull())
+  void TraverseTypeHelper(QualType qual_type) {
+    if (qual_type.isNull())
       return;
 
-    if (const auto* template_spec = type->getAs<TemplateSpecializationType>()) {
-      for (const TemplateArgument& arg : template_spec->template_arguments())
-        TraverseTemplateArgument(arg);
-    }
+    const Type* const type = qual_type.getTypePtr();
+    const Type* const desugared_until_alias_or_tpl = Desugar(type);
+    seen_types_.insert(desugared_until_alias_or_tpl);
+
+    const Type* const desugared =
+        DesugarAliasTypes(desugared_until_alias_or_tpl);
+    // A desugared type shouldn't be inserted into resugar_map along with its
+    // sugared version as a value instead of a key, hence it is important to
+    // call the base class implementation of TraverseType and not the own.
+    // TODO(bolshakov): check whether typedef components are "provided".
+    if (type != desugared)
+      Base::TraverseType(QualType{desugared, 0});
```

Comment:
Please avoid initialization syntax, I find it's just a distraction -- `QualType(desugared, 0)` means the same in this context.

Reply:
Ok, but the "C++ Core Guidelines" [suggest the opposite](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es64-use-the-tenotation-for-construction).

Reply:
This two-step recursion makes my head spin a little. Does this alter the traversal so that for every type we also traverse its fully desugared type? I'm trying to visualize what that looks like, but coming up short. Is there a reasonably compact C++ example to demonstrate what is picked up, that you could put in a doc comment for this function?

I'm a little thrown by the `Base::TraverseType` here, and then another `Base::TraverseType` in the calling `TraverseType` function, but I guess they traverse the desugared and sugared types respectively?

Could this be called `TraverseDesugared` instead?

Reply:
And the type is not collected into `seen_types_` in `VisitType` because... you would have to desugar and only report it if it's different there too? It feels like the type collection would be more suitable for `Visit`, whereas traversal adjustments belong in `Traverse`.

But these are mostly questions, not suggestions.

Reply:
I suspect that `if (type != desugared)` could be removed along with `Base::TraverseType*` calls inside `Base::TraverseType` and `Base::TraverseTypeLoc` methods. I had just faced with infinite recursion during the work, but I hope that situation won't repeat anymore in the last version...

Reply:
If `VisitType` filled in `seen_types_`, then `desugared` would be inserted into `seen_types_` due to `VisitType` call from inside `Base::TraverseType(QualType{desugared, 0});` call, which should be avoided, as I mentioned before.